### PR TITLE
Editor: support for custom meta boxes

### LINKED ISSUE #3134: Support Legacy WordPress Meta Boxes
## Feature Description

In https://github.com/Yoast/wordpress-seo/pull/15532 / #113 we made the first step to support plugins like Yoast SEO.

To give users control over things like meta descriptions and Open Graph descriptions, we need to support WordPress' legacy meta box system, so that we can display meta boxes like the one from Yoast SEO.

There's a lot of prior art for this in WordPress and Gutenberg.

## Additional Context

Yoast SEO support was mentioned in support requests like #3086 and #3107.

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

This PR lays the foundation for supporting custom meta boxes as envisioned in #3134, by closely following the precedence set by Gutenberg.

Thankfully, much of the required logic is provided by WordPress core.

**How it works in a nutshell**

1. Meta boxes are rendered in `edit-story.php`
1. The `MetaBoxes` and `MetaBoxesArea` components moves the meta boxes to the right place in the DOM, and ensure expected behavior like collapsing works -> just like Gutenberg does it
1. `useSaveMetaBoxes` gets the data from all the meta boxes and prepares it for sending to WordPress
1. Data is sent to the pre-existing meta boxes endpoint in core

WordPress really does all the heavy lifting.

All that was needed on our side is following Gutenberg's lead and load the meta boxes in the same way

## Relevant Technical Choices

* Cleaned up and unified CSS between dashboard and editor.
* New `Story_Post_Type::get_meta_boxes_per_location()` method so the editor knows which locations are empty and don't need to be displayed.
* New `Story_Post_Type::remove_meta_boxes()` method to remove meta boxes that don't support the block editor.  
  Mimics what `do_meta_boxes()` does for the block editor already. Since we're not a block editor, we have to do this manually.
* New `MetaBoxes` component that renders `MetaBoxesArea` for a couple of locations (same 2 locations Gutenberg supports)  
  Does not render anything if feature flag is disabled.
  Also makes sure meta boxes can be toggled/collapsed.
  Inspired by https://github.com/WordPress/gutenberg/blob/78585d6935fee9020017d17383cef597b67c5703/packages/edit-post/src/components/meta-boxes/index.js
* New `MetaBoxesArea` component that displays a single meta box area.   
  Moves the metabox sub-tree to be within our app.
  Inspired by https://github.com/WordPress/gutenberg/blob/78585d6935fee9020017d17383cef597b67c5703/packages/edit-post/src/components/meta-boxes/meta-boxes-area/index.js
* In `MetaBoxesarea`, slightly override default meta box styling  
  Inspired by https://github.com/WordPress/gutenberg/blob/78585d6935fee9020017d17383cef597b67c5703/packages/edit-post/src/components/meta-boxes/meta-boxes-area/style.scss
* Renders `MetaBoxes` in `Layout` so it's better detached from the canvas and can be more easily refactored / extracted in the future.
* New `useSaveMetaBoxes` hook that does the actual saving of metabox data.  
  Takes the form data and sends it to the meta boxes endpoint in WordPress
  Important: it only saves meta boxes when we just finished saving the story, but not if that save was an autosave.
  Inspired by https://github.com/WordPress/gutenberg/blob/148e2b28d4cdd4465c4fe68d97fcee154a6b209a/packages/edit-post/src/store/effects.js#L24-L126

**Things _not_ implemented**

* Allowing users to show/hide certain meta boxes.  
  Don't think we need this, doesn't seem common.
  Gutenberg does it like this: https://github.com/WordPress/gutenberg/blob/148e2b28d4cdd4465c4fe68d97fcee154a6b209a/packages/edit-post/src/store/selectors.js#L211-L304
* No support for the "Custom Fields" meta box.  
  This one is rarely needed and also quite complex — even Gutenberg makes this opt-in only. Don't think we should support this.

## To-do

* [x] Add tests where possible and reasonable.
* [x] Manual testing

Follow-up tickets:

* [ ] Get UX feedback, improve styling, etc.

## User-facing changes

Experiment disabled (no change):

Screenshot 2020-11-04 at 11 41 08

Experiment enabled:

All in One SEO Pack enabled, but collapsed:

Screenshot 2020-11-04 at 11 42 03

All in One SEO Pack enabled, not collapsed:

Screenshot 2020-11-04 at 11 42 12

### Default state

1. Verify that editor looks the same as always.
1. Enable meta box feature flag on Experiments page
1. Verify that editor looks the same as always.

### Yoast SEO

**Note:** The Yoast SEO meta box does not yet seem to work on WordPress 5.6, not even in Gutenberg. So don't test this on 5.6.

**Second Note:** For testing Yoast SEO, check out this PR: https://github.com/Yoast/wordpress-seo/pull/16298

1. Install Yoast SEO
1. Enable meta box feature flag on Experiments page
1. Create a new story
1. Verify that Yoast SEO meta box is displayed and can be collapsed
1. Save story, reload page
1. Verify that everything was saved as expected

### All in One SEO Pack

1. Install All in One SEO pack
1. In the plugin's settings, enable support for the Web Stories post type.
1. Enable meta box feature flag on Experiments page
1. Create a new story
1. Verify that AiO meta box is displayed and can be collapsed
1. Edit content in meta box
1. Save story, reload page
1. Verify that everything was saved as expected

---

Fixes #3134

### CODE REVIEW

File: assets/src/dashboard/style.css [hunk 1]
```diff
@@ -75,7 +74,3 @@ body.web-story_page_stories-dashboard #wpfooter {
 button {
   font-size: inherit;
 }
-
```
Comment (spacedmonkey) [inline assets/src/dashboard/style.css:78]: Why was this removed?

File: assets/src/edit-story/style.css [hunk 2]
```diff
@@ -48,17 +52,8 @@ body.edit-story #wpfooter {
   display: none;
 }
 
-body.edit-story #edit-story .components-spinner {
```
Comment (spacedmonkey) [inline assets/src/edit-story/style.css:51]: Why was this removed?

File: assets/src/dashboard/style.css (hunk 1 above)
Comment (swissspidy) [reply assets/src/dashboard/style.css:78]: Not removed per se, but I restructured the CSS a bit. If you look at Percy the visual outcome should be the same.

File: assets/src/edit-story/style.css (hunk 2 above)
Comment (swissspidy) [reply assets/src/edit-story/style.css:51]: We don't use `components-spinner` anywhere

File: includes/Story_Post_Type.php [hunk 3]
```diff
@@ -610,6 +613,106 @@ public function filter_use_block_editor_for_post_type( $use_block_editor, $post_
...
+		// We don't currently support the 'Custom Fields' meta box.
+		remove_meta_box( 'postcustom', $screen, $screen->id );
+
+		foreach ( [ 'normal', 'advanced' ] as $context ) {
+			if ( ! isset( $wp_meta_boxes[ $screen->id ][ $context ] ) ) {
+				continue;
+			}
+
+			foreach ( [ 'high', 'sorted', 'core', 'default', 'low' ] as $priority ) {
+				foreach ( (array) $wp_meta_boxes[ $screen->id ][ $context ][ $priority ] as $meta_box ) {
```
Comment (spacedmonkey) [inline includes/Story_Post_Type.php]:
```suggestion
                                 if ( !isset( $wp_meta_boxes[ $screen->id ][ $context ][ $priority ] )  ){
                                       continue;
                                 }
                                  
				foreach ( (array) $wp_meta_boxes[ $screen->id ][ $context ][ $priority ] as $meta_box ) {
```

This fixed a notice warning I got locally.

**Comment by spacedmonkey** [comment]
image
This PR didn't work for me and notice error. I tihnk it makes sense to get this to working state before handing it off.

**Comment by swissspidy** [comment]
@spacedmonkey As I said yesterday reach out before testing with Yoast SEO. Also, Yoast SEO doesn‘t work with 5.6 beta. Also, I will write a summary before handing over, as discussed.

**Comment by spacedmonkey** [comment]
> @spacedmonkey As I said yesterday reach out before testing with Yoast SEO. Also, Yoast SEO doesn‘t work with 5.6 beta. Also, I will write a summary before handing over, as discussed.

Not going to lie, completely forgot you said this. I await your notes on this.

File: assets/src/edit-story/app/api/apiProvider.js [hunk 4]
```diff
@@ -230,6 +230,29 @@ function APIProvider({ children }) {
     return apiFetch({ path: addQueryArgs(users, { per_page: '-1' }) });
   }, [users]);
 
+  const saveMetaBoxes = useCallback(
```
Comment (swissspidy) [inline assets/src/edit-story/app/api/apiProvider.js:252]: This callback is inspired by this section in Gutenberg: https://github.com/WordPress/gutenberg/blob/148e2b28d4cdd4465c4fe68d97fcee154a6b209a/packages/edit-post/src/store/effects.js#L72-L126

File: assets/src/edit-story/app/layout/layout.js [hunk 5]
```diff
@@ -66,6 +68,7 @@ const Area = styled.div`
   z-index: 2;
 `;
 
+// TODO: Fix meta boxes layout.
```
Comment (swissspidy) [inline assets/src/edit-story/app/layout/layout.js:71]: I think the current grid layout is _OK_. Can ask Sam in follow-up tickets for UX.

File: assets/src/edit-story/components/metaBoxes/metaBoxesArea.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+`;
+
+const Spinner = styled.div`
+  position: absolute;
+  top: 10px;
+  right: 100px;
+  z-index: 1;
+`;
+
+function MetaBoxesArea({ location }) {
```
Comment (swissspidy) [inline assets/src/edit-story/components/metaBoxes/metaBoxesArea.js]: This component is basically https://github.com/WordPress/gutenberg/blob/78585d6935fee9020017d17383cef597b67c5703/packages/edit-post/src/components/meta-boxes/meta-boxes-area/index.js, but a bit simpler and more modern

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+function useSaveMetaBoxes({ story, isSaving, isAutoSaving }) {
```
Comment (swissspidy) [inline assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js:41]: This effect hook is inspired by https://github.com/WordPress/gutenberg/blob/148e2b28d4cdd4465c4fe68d97fcee154a6b209a/packages/edit-post/src/store/effects.js#L24-L126

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    if (
+      !isFeatureEnabled ||
+      isSaving ||
+      isAutoSaving ||
+      isSavingMetaBoxes ||
+      !wasSaving ||
+      wasAutoSaving
+    ) {
+      return;
+    }
```
Comment (swissspidy) [inline assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: This is key:

We only want to save meta boxes when we _just_ finished saving the story, but not if that save was an autosave.

File: assets/src/edit-story/app/story/storyProvider.js [hunk 9]
```diff
@@ -154,7 +162,7 @@ function StoryProvider({ storyId, children }) {
       animationState,
       capabilities,
       meta: {
-        isSaving: isSaving || isAutoSaving,
+        isSaving: isSaving || isAutoSaving || isSavingMetaBoxes,
```
Comment (swissspidy) [inline assets/src/edit-story/app/story/storyProvider.js:165]: This is pretty cool:

The user does not realize we're first saving the story and then the meta boxes. They just see one continuous saving animation.

**Comment by swissspidy** [comment]
@spacedmonkey Heya, I took some time to update the PR description to better describe what this does and what's left to do.

On the surface it's pretty straightforward: print only relevant meta boxes -> save meta boxes like Gutenberg does -> let WordPress do the heavy lifting

PTAL and ask questions for anything that's unclear.

Overall I think this is actually in a pretty good shape already (for a feature-flagged enhancement)

**Comment by swissspidy** [comment]
@miina @barklund requested your review here due to changes to the story provider and workspace layout

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 8 above)
Comment (miina) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Should this only be triggered if there are any metaboxes to save? Also, is it always saving even if there are no actual changes? Could you confirm this is not influencing history?

File: assets/src/edit-story/components/metaBoxes/index.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
+/*
```
Comment (miina) [inline assets/src/edit-story/components/metaBoxes/index.js]: I wonder if it would be good to create an integrations folder to separate the WP-specific logic. WDYT?

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 8 above)
Comment (swissspidy) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: > Should this only be triggered if there are any metaboxes to save?
Good point, definitely should only trigger this if there are actually meta boxes. Will add this.

> Also, is it always saving even if there are no actual changes?
That's right, we don't (and can't really) detect whether there were any changes to the meta boxes. We just send the update and WordPress will figure it out server-side.

> Could you confirm this is not influencing history?

I'll double check, but this should have no impact on history at all, meta box changes are not added to history.

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 7 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js:41]: Worth adding this a jsdoc block?

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  const area = document.querySelector(
+    `.web-stories-meta-boxes-area-${location} .metabox-location-${location}`
+  );
+
+  return (
+    area || document.querySelector('#metaboxes .metabox-location-' + location)
+  );
+}
+
+function usePrevious(value) {
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: JSdoc block to explain what this does.

File: assets/src/edit-story/app/api/apiProvider.js (hunk 4 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/api/apiProvider.js:252]: This is worth a comment in the code as well.

File: assets/src/edit-story/app/story/storyProvider.js (hunk 9 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/story/storyProvider.js:165]: In my testing, saving a story now takes around 5 seconds. Is this normal? As much as I think this is cool, I think we should maybe proritize the story saving. I think that is what important, if the meta box fails to save, then so be it.

File: includes/templates/admin/dashboard.php [hunk 12]
```diff
@@ -29,10 +29,19 @@
...
+		<h1 class="loading-message"><?php esc_html_e( 'Please wait...', 'web-stories' ); ?></h1>
+	</div>
+
+	<?php // JavaScript is disabled. ?>
+	<div class="wrap hide-if-js web-stories-wp-no-js">
+		<h1 class="wp-heading-inline"><?php esc_html_e( 'Web Stories', 'web-stories' ); ?></h1>
+		<div class="notice notice-error notice-alt">
+			<p><?php esc_html_e( 'Web Stories for WordPress requires JavaScript. Please enable JavaScript in your browser settings.', 'web-stories' ); ?></p>
+		</div>
+	</div>
```
Comment (spacedmonkey) [inline includes/templates/admin/dashboard.php:46]: This is repeated isn't it?

Couldn't we just put this in `error-no-js.php` like it was before.

File: assets/src/edit-story/integrations/wordpress/components/metaBoxes/index.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+ */
+import { useConfig } from '../../../../app/config';
+import MetaBoxesArea from './metaBoxesArea';
+
+const Wrapper = styled.div``;
+
+function MetaBoxes() {
+  const customMetaBoxes = useFeature('customMetaBoxes');
+
+  const { postType, metaBoxes } = useConfig();
```
Comment (swissspidy) [inline assets/src/edit-story/integrations/wordpress/components/metaBoxes/index.js]:
```suggestion
  const { postType, metaBoxes = {} } = useConfig();
```

File: assets/src/edit-story/integrations/wordpress/components/metaBoxes/index.js [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  if (!customMetaBoxes) {
+    return null;
+  }
+
+  const locations = ['normal', 'advanced'];
+
+  return (
+    <Wrapper>
+      {locations.map((location) => {
+        const isVisible = Boolean(metaBoxes[location].length);
```
Comment (swissspidy) [inline assets/src/edit-story/integrations/wordpress/components/metaBoxes/index.js]:
```suggestion
        const isVisible = Boolean(metaBoxes[location]?.length);
```

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+  return ref.current;
+}
+
+function useSaveMetaBoxes({ story, isSaving, isAutoSaving }) {
+  const isFeatureEnabled = useFeature('customMetaBoxes');
+  const { metaBoxes = {} } = useConfig();
+
+  const locations = ['normal', 'advanced'];
+  const hasMetaBoxes = locations.some((location) =>
+    Boolean(metaBoxes[location].length)
```
Comment (swissspidy) [inline assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]:
```suggestion
    Boolean(metaBoxes[location]?.length)
```

File: assets/src/edit-story/app/api/apiProvider.js [hunk 16]
```diff
@@ -239,6 +239,29 @@ function APIProvider({ children }) {
...
+    (story, formData) => {
+      // Additional data needed for backward compatibility.
+      // If we do not provide this data, the post will be overridden with the default values.
+      const additionalData = [
+        story.comment_status ? ['comment_status', story.comment_status] : false,
+        story.ping_status ? ['ping_status', story.ping_status] : false,
+        story.sticky ? ['sticky', story.sticky] : false,
+        story.author ? ['post_author', story.author] : false,
```
Comment (swissspidy) [inline assets/src/edit-story/app/api/apiProvider.js:261]: Need `story.author.id` soon?

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 8 above)
Comment (miina) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Hmm, if it doesn't influence the history at all, then I wonder how can the user save only the updates from the meta-boxes in Draft mode? E.g. the user has no new changes and then changes the content of the meta boxes. Since history doesn't detect any new changes, the Draft button stays disabled. Thoughts?

File: assets/src/edit-story/app/story/storyProvider.js (hunk 9 above)
Comment (swissspidy) [reply assets/src/edit-story/app/story/storyProvider.js:165]: Note that saving the story itself is not impacted, it's just that the saving _animation_ is displayed a bit longer.

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+      // Saves the wp_editor fields.
+      global.tinyMCE?.triggerSave();
+
+      // We gather all the metaboxes locations data and the base form data.
+      const baseFormElement = document.querySelector('.metabox-base-form');
+      const baseFormData = new global.FormData(baseFormElement || undefined);
+
+      const formDataToMerge = [
+        baseFormData,
+        ...['normal', 'advanced'].map(
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Isn't this a repeat of ` const locations `?

**Comment by spacedmonkey** [comment]
Screenshot 2020-11-17 at 12 23 27
Gutenberg has these options to hide and show metaboxes. I am guessing we don't need this here.

**Comment by spacedmonkey** [comment]
Screenshot 2020-11-17 at 12 36 08
In my testing, I adding some custom fields in ACF. The pages are no longer visual.

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 8 above)
Comment (miina) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: ^ @swissspidy Thoughts on this in case you missed it among the notifications?

File: includes/Story_Post_Type.php [hunk 18]
```diff
@@ -414,6 +420,113 @@ public function filter_use_block_editor_for_post_type( $use_block_editor, $post_
...
+
+		if ( ! $screen instanceof WP_Screen ) {
+			return;
+		}
+
+		// We don't currently support the 'Custom Fields' meta box.
+		remove_meta_box( 'postcustom', $screen, $screen->id );
+
+		$locations  = [ 'side', 'normal', 'advanced' ];
+		$priorities = [ 'high', 'sorted', 'core', 'default', 'low' ];
```
Comment (spacedmonkey) [inline includes/Story_Post_Type.php]: This is repeated, should be a class level property?

File: includes/Story_Post_Type.php [hunk 19]
```diff
@@ -524,12 +655,14 @@ public function get_editor_settings() {
...
 					'link'        => '/web-stories/v1/link',
 					'statusCheck' => '/web-stories/v1/status-check',
+					'metaboxes'   => $meta_box_url,
 				],
 				'metadata'         => [
 					'publisher' => $this->get_publisher_data(),
 				],
 				'version'          => WEBSTORIES_VERSION,
 				'encodeMarkup'     => $this->decoder->supports_decoding(),
+				'metaBoxes'        => $this->get_meta_boxes_per_location(),
```
Comment (spacedmonkey) [inline includes/Story_Post_Type.php]: I don't think we should add this unless this is feature flag is enabled. We could do this 
```
if ( $this->experiments->is_experiment_enabled( 'customMetaBoxes' ) ) {
     $settings['config']['metaBoxes'] = $this->get_meta_boxes_per_location();
     // Get admin url for handling meta boxes.
     $meta_box_url = admin_url( 'post.php' );
     $meta_box_url = add_query_arg(
          [
              'post'                  => $story_id,
              'action'                => 'edit',
              'meta-box-loader'       => true,
              'meta-box-loader-nonce' => wp_create_nonce( 'meta-box-loader' ),
          ],
          $meta_box_url
     );
     $settings['api']['metaboxes'] = $meta_box_url;
}
```

File: includes/templates/admin/edit-story.php [hunk 20]
```diff
@@ -29,7 +29,7 @@
 	die( '-1' );
 }
 
-global $post_type, $post_type_object, $post;
+global $post_type, $post_type_object, $post, $wp_meta_boxes;
```
Comment (spacedmonkey) [inline includes/templates/admin/edit-story.php]: Why is this global needed? `$wp_meta_boxes`

File: includes/Story_Post_Type.php [hunk 21]
```diff
@@ -414,6 +420,113 @@ public function filter_use_block_editor_for_post_type( $use_block_editor, $post_
...
+	 *
+	 * Used to disable empty meta boxes in the editor.
+	 *
+	 * @see the_block_editor_meta_boxes()
+	 *
+	 * @since 1.2.0
+	 *
+	 * @return array List of meta boxes per location.
+	 */
+	public function get_meta_boxes_per_location() {
```
Comment (spacedmonkey) [inline includes/Story_Post_Type.php]: This class is really very long. Should this logic live in a metabox class of it's own? 
```
$metabox = new Metaboxes( Story_Post_Type::POST_TYPE_SLUG );
$metabox->init();
```

 Or an abstract class called `Post_Type`, that both `Story_Post_Type` and `Template_Post_Type` extend. That way we can keep much of the shared logic between post types shared.

File: includes/Story_Post_Type.php (hunk 18 above)
Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: Also https://github.com/google/web-stories-wp/pull/5048/files#r525146639

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 17 above)
Comment (swissspidy) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Yep, good catch!

File: includes/Story_Post_Type.php (hunk 19 above)
Comment (swissspidy) [reply includes/Story_Post_Type.php]: Seems unnecessary. In the editor it's a no-op anyway if the feature is disabled.

File: includes/templates/admin/edit-story.php (hunk 20 above)
Comment (swissspidy) [reply includes/templates/admin/edit-story.php]: Hmm I'll check, might be unneeded.

**Comment by swissspidy** [comment]
> Gutenberg has these options to hide and show metaboxes. I am guessing we don't need this here.

Correct. At least not for now.

> In my testing, I adding some custom fields in ACF. The pages are no longer visual.

Need to collapse the meta box again to see the editor. It's a limitation of the current CSS grid layout which we can improve in further PRs.

File: includes/Story_Post_Type.php (hunk 21 above)
Comment (swissspidy) [reply includes/Story_Post_Type.php]: > This class is really very long. Should this logic live in a metabox class of its own?

I'll take a stab at that.
 
> Or an abstract class called `Post_Type`, that both `Story_Post_Type` and `Template_Post_Type` extend. That way we can keep much of the shared logic between post types shared.

That's something that's better explored & discussed at #5062

Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: There is no need to make this class longer and make the problem worse 😛

File: includes/Story_Post_Type.php (hunk 19 above)
Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: Why run the methods when they are not used?

Comment (spacedmonkey) [reply includes/Story_Post_Type.php]: It also hide the feature completely unless the flag is enabled.

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 8 above)
Comment (swissspidy) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: So the way Gutenberg does it is by forcing  the "Save" button to be enabled if there are any active meta boxes:

https://github.com/WordPress/gutenberg/blob/f4300ff87ea9a8e5d2aed155f180c76b83a54c4a/packages/edit-post/src/components/header/index.js#L46-L58

That could be a viable option for us as well, because trying to add meta boxes to history or generally detecting any changes in meta boxes is going to be very complex and most probably not worth trying.

Comment (miina) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Makes sense to me!

**Comment by swissspidy** [comment]
@spacedmonkey Since I am out, feel free to do that `MetaBox` PHP class refactoring you suggested here if you'd like.

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 17 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: ⬆️

File: assets/src/edit-story/components/header/buttons/update.js [hunk 22]
```diff
@@ -55,25 +68,24 @@ function Update() {
...
+      >
+        {__('Save draft', 'web-stories')}
+      </Outline>
+    );
+  }
+
+  let text = __('Update', 'web-stories');
+
+  if (status === 'future') {
+    text = __('Schedule', 'web-stories');
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/header/buttons/update.js]: @swissspidy Can you explain change a little. The switch statement before was much easier to read than the way it has been changed.

File: includes/Meta_Boxes.php [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+
+	/**
+	 * Init.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @return void
+	 */
+	public function init() {
+		add_action( 'add_meta_boxes_' . Story_Post_Type::POST_TYPE_SLUG, [ $this, 'remove_meta_boxes' ], PHP_INT_MAX );
```
Comment (spacedmonkey) [inline includes/Meta_Boxes.php:57]: This feel odd. Can we make it is class property like this. 

```
add_action( 'add_meta_boxes_' . $this->post_type, [ $this, 'remove_meta_boxes' ], PHP_INT_MAX );
```

```
$this->meta_boxes = new Meta_Boxes( Story_Post_Type::POST_TYPE_SLUG );
```


Meaning in the future / in tests we could add another CPT.

File: assets/src/edit-story/components/header/buttons/update.js (hunk 22 above)
Comment (swissspidy) [reply assets/src/edit-story/components/header/buttons/update.js]: I found it easier this way. Was odd to only return something in the default case, and there were only 2 other cases. Now it simply returns early when possible.

File: includes/Meta_Boxes.php (hunk 23 above)
Comment (spacedmonkey) [reply includes/Meta_Boxes.php:57]: Or better yet, this line does live here at all. 

In  Story_Post_Type::init 

```
add_action( 'add_meta_boxes_' . self::POST_TYPE_SLUG, [ $this->meta_box, 'remove_meta_boxes' ], PHP_INT_MAX );
```

Comment (swissspidy) [reply includes/Meta_Boxes.php:57]: Found it unnecessary, but sure we can do that.

File: tests/phpunit/tests/Meta_Boxes.php [hunk 24]
```diff
@@ -23,4 +15,18 @@
  * limitations under the License.
  */
 
-// Silence is golden.
+namespace Google\Web_Stories\Tests;
+
+/**
+ * @coversDefaultClass \Google\Web_Stories\Meta_Boxes
+ */
+class Meta_Boxes extends \WP_UnitTestCase {
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Meta_Boxes.php:25]: Can't we have some tests here?

File: includes/Meta_Boxes.php (hunk 23 above)
Comment (swissspidy) [reply includes/Meta_Boxes.php:57]: > Or better yet, this line does live here at all.
> 
> In Story_Post_Type::init
> 
> ```
> add_action( 'add_meta_boxes_' . self::POST_TYPE_SLUG, [ $this->meta_box, 'remove_meta_boxes' ], PHP_INT_MAX );
> ```

I see your point, but it's quite different from all our other classes, no? All other classes register their hooks in their `init()` method.

(Aside: If we can't agree on the best way right away, happy to open a dedicated ticket to address this. I really wanna merge it soon :)

File: includes/Story_Post_Type.php [hunk 25]
```diff
@@ -524,12 +546,14 @@ public function get_editor_settings() {
...
 					'link'        => '/web-stories/v1/link/',
 					'statusCheck' => '/web-stories/v1/status-check/',
+					'metaBoxes'   => $this->meta_boxes->get_meta_box_url( (int) $story_id ),
 				],
 				'metadata'         => [
 					'publisher' => $this->get_publisher_data(),
 				],
 				'version'          => WEBSTORIES_VERSION,
 				'encodeMarkup'     => $this->decoder->supports_decoding(),
+				'metaBoxes'        => $this->meta_boxes->get_meta_boxes_per_location(),
```
Comment (spacedmonkey) [inline includes/Story_Post_Type.php:556]: Neat 👍

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+ * @param {Object} props.story Story object.
+ * @param {boolean} props.isSaving Whether saving is in progress.
+ * @param {boolean} props.isAutoSaving Whether autosaving is in progress.
+ * @return {{isSavingMetaBoxes: boolean}} Metaboxes status.
+ */
+function useSaveMetaBoxes({ story, isSaving, isAutoSaving }) {
+  const isFeatureEnabled = useFeature('customMetaBoxes');
+  const { metaBoxes = {} } = useConfig();
+
+  const locations = ['normal', 'advanced'];
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Should we pass this in config, from php to js? This variable is repeated in both PHP and js.

File: assets/src/edit-story/integrations/wordpress/components/metaBoxes/metaBoxesArea.js [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+function MetaBoxesArea({ location }) {
+  const formRef = useRef();
+  const containerRef = useRef();
+
+  const { isSaving } = useStory(({ state }) => ({
+    isSaving: state.meta.isSaving,
+  }));
+
+  useEffect(() => {
+    const form = document.querySelector(`.metabox-location-${location}`);
```
Comment (spacedmonkey) [inline assets/src/edit-story/integrations/wordpress/components/metaBoxes/metaBoxesArea.js:115]: How is this different from `getMetaBoxContainer`?

File: includes/Meta_Boxes.php (hunk 23 above)
Comment (spacedmonkey) [reply includes/Meta_Boxes.php:57]: We could convert this into a trait and just use this 
```
add_action( 'add_meta_boxes_' . self::POST_TYPE_SLUG, [ $this, 'remove_meta_boxes' ], PHP_INT_MAX );
```

File: assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js (hunk 26 above)
Comment (swissspidy) [reply assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js]: Can just do `Object.keys(metaBoxes)` I guess

File: assets/src/edit-story/integrations/wordpress/components/metaBoxes/metaBoxesArea.js (hunk 27 above)
Comment (swissspidy) [reply assets/src/edit-story/integrations/wordpress/components/metaBoxes/metaBoxesArea.js:115]: This effect does:

Find meta box form container on the page, and move it to a different DOM sub tree. (Mostly for styling)

`getMetaBoxContainer` does:

Find meta box form container on the page, regardless of whether it was already moved or not.

----

Now that I am writing this: perhaps there's no need for `getMetaBoxContainer` at all. Could just do `document.querySelector(``.metabox-location-${location}``)` in both places. Will test.

File: assets/src/edit-story/components/header/buttons/update.js (hunk 22 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/header/buttons/update.js]: Getting rid of multiple if statement to make them more readable is exactly the point of a switch statement. If this refactor is not requires for this functionality it should be removed and handled in another PR.

File: includes/Meta_Boxes.php (hunk 23 above)
Comment (swissspidy) [reply includes/Meta_Boxes.php:57]: I don't think a trait is a good solution for this kinda scenario

**Review by spacedmonkey — APPROVED** [review]
👍 Very happy to approve this!
